src/i_video.c

```diff
@@ -1389,15 +1408,59 @@ static void CreateUpscaledTexture(boolean force)
     SDL_SetTextureScaleMode(texture_upscaled, SDL_ScaleModeLinear);
 }
 
-static void ResetLogicalSize(void)
+static void UpdateMarginState(void)
 {
-    blit_rect.w = video.width;
-    blit_rect.h = video.height;
+    int w = 1, h = 1;
+    SDL_GetRendererOutputSize(renderer, &w, &h);
+
+    float scalex, scaley;
+    SDL_RenderGetScale(renderer, &scalex, &scaley);
+    SDL_RenderSetScale(renderer, 1.f, 1.f);
 
-    if (SDL_RenderSetLogicalSize(renderer, video.width, actualheight))
+    SDL_Rect vp;
+    SDL_RenderGetViewport(renderer, &vp);
+    has_margins = vp.x != 0 || vp.y != 0 || vp.w != w || vp.h != h;
+
+    SDL_RenderSetScale(renderer, scalex, scaley);
+}
+
+static void UpdateViewport(void)
+{
+    int w, h;
+
+     // nocheckin
+    boolean prevmargins = has_margins;
+
+    if (full_frame
+        && SDL_GetRendererOutputSize(renderer, &w, &h) == 0
+        && fabs((double)w / h - CurrentAspectRatio()) < .002)
+    {
+        float scalex = (float)w / video.width;
+        float scaley = (float)h / actualheight;
+        SDL_RenderSetLogicalSize(renderer, 0, 0);
```

Comment:
Can we remove `SDL_RenderSetLogicalSize` entirely and just use `SDL_RenderSetScale`?

Reply:
The viewport needs to be reset as well or it could be left with the wrong settings, see https://github.com/libsdl-org/SDL/blob/30972d7e7fce4430bdc911926811d968eec7af57/src/render/SDL_render.c#L2438

Reply:
I mean we can just not call `SDL_RenderSetLogicalSize` at all and replace it with your `UpdateViewport` function.

Reply:
It still does some stuff that is nice to not have to write ourselves, keeping aspect ratio and centering the viewport for the cases where we want that in pillar/letterbox, translating coords of mouse events, etc.

Reply:
Please try #1847, I'm not sure if it fixes the "margins" issue. Replacing `SDL_RenderSetLogicalSize` is easy, seems to work for me.